zcash_primitives/src/transaction/builder.rs

```diff
@@ -18,56 +19,75 @@ use crate::{
     sapling::{prover::TxProver, Diversifier, Node, Note, PaymentAddress},
     transaction::{
         components::{
-            amount::{Amount, DEFAULT_FEE},
+            amount::{Amount, BalanceError},
             sapling::{
                 self,
-                builder::{SaplingBuilder, SaplingMetadata},
+                builder::{SaplingBuilder, SaplingInput, SaplingMetadata, SaplingOutput},
+            },
+            transparent::{
+                self,
+                builder::{TransparentBuilder, TransparentInput},
             },
-            transparent::{self, builder::TransparentBuilder},
         },
+        fees::{FeeRule, Infallible},
         sighash::{signature_hash, SignableInput},
         txid::TxIdDigester,
-        Transaction, TransactionData, TxVersion, Unauthorized,
+        Transaction, TransactionData, TxOut, TxVersion, Unauthorized,
     },
     zip32::ExtendedSpendingKey,
 };
 
-#[cfg(feature = "transparent-inputs")]
-use crate::transaction::components::transparent::TxOut;
-
 #[cfg(feature = "zfuture")]
 use crate::{
     extensions::transparent::{ExtensionTxBuilder, ToPayload},
-    transaction::components::{
-        tze::builder::TzeBuilder,
-        tze::{self, TzeOut},
+    transaction::{
+        components::{
+            tze::builder::TzeBuilder,
+            tze::{self, TzeOut},
+        },
+        fees::FutureFeeRule,
     },
 };
 
-#[cfg(any(test, feature = "test-dependencies"))]
-use crate::sapling::prover::mock::MockTxProver;
-
 const DEFAULT_TX_EXPIRY_DELTA: u32 = 20;
 
+/// Errors that can occur during transaction construction.
 #[derive(Debug, PartialEq, Eq)]
-pub enum Error {
-    ChangeIsNegative(Amount),
-    InvalidAmount,
-    NoChangeAddress,
+pub enum Error<FeeError> {
+    /// Insufficient funds were provided to the transaction builder; the given
+    /// additional amount is required in order to construct the transaction.
+    InsufficientFunds(Amount),
```

Comment:
This will surely be a UX problem.  Is there a way of doing a `dry-run` to know the estimated fee for a given amount to show on wallets' send flow? Waiting for the user to complete the whole transaction to deny it because of variable fees not calculated well beforehand seems pretty inconvenient

Reply:
Usually insufficient funds will be detected at note selection time, so the build-time check is really just a fall-back to ensure that the transaction doesn't violate balance rules. Also, the next generation of transaction construction will provide an intermediate result that has the privacy implications of the transaction listed out for the user's approval.

Reply:
> Also, the next generation of transaction construction will provide an intermediate result that has the privacy implications of the transaction listed out for the user's approval.

I'm not following you here. In terms of wallet UX, can we have a function that given an amount, it will return an approximate fee so that wallet developers can validate their "send funds" forms UI?

Reply:
There's a subtlety here which is that we would need such a function to essentially prepare the transaction and "reserve" the notes/UTXOs that would be spent, so that we can be sure that the fee doesn't change between the time that the UI displays the fee and the time that the user goes to actually confirm the transaction; for example, if a new note comes in that would be preferentially selected by the note selection algorithm, the fees and the notes being spent could change in that interim. So, what we need to do is what we've discussed before: prepare the transaction and lock the notes to be spent to it, and then return that prepared transaction (along with its privacy implications) to be confirmed by the user. The fee should be displayed on the confirmation page.

Reply:
The problem with this approach is that it doesn't solve the problem I'm referring to, which is: "don't let the user proceed to a place that inevitably is an error state if it can be known beforehand". 

The SDK will update every 20~ seconds and being the user in the send flow, it will refresh the available information and that information in the local state will be used to create the transaction. Any changes in balance should trigger a UI refresh in send flow so that it can be validated against the latest current state.

Reply:
We do know that there are certain combinations of amount and notes used that will make the fees an impediment for the transaction to be sent. The "dry-run" function I'm proposing is aimed to catch and report that particular situation so that wallet developers can validate their send form before proceeding. There will be situations like you mention where timing and chance will plot against it. *

Also, it can be the case that a wallet developer does not want a confirmation screen. So this additional step to safety check will not be present. Which makes this validation function more helpful to avoid failed transactions for consensus reasons which are a pain to explain to users.




* I hadn't thought of this. Can users be dusted into oblivion thanks to these new fee scheme?

Reply:
No; it is the responsibility of the ZIP 321 note selection implementation to ignore notes that are below the value of the fee required to spend them (though low-valued notes can have their value recovered by using them as a free input when spending a note that is by itself large enough to cover the full transaction amount.)

Reply:
In any case, this error needs to exist because we must *always* validate the transaction at construction time anyway. The transaction preparation step you're talking about is effectively the input selection step itself; it's been renamed 'propose_transaction' to make this a bit clearer: https://github.com/zcash/librustzcash/pull/689/files#diff-fedecff319cb43f7a4f8b913cdee2332ba452a6310e2470cf6cc7631ca83291cR126